ARROW-12468: [Python][R] Expose ScannerBuilder::UseAsync to Python & R

westonpace opened the pull request:
This exposes (in python & R) the `UseAsync` method added to `ScannerBuilder` in ARROW-12289.

lidavidm commented on python/pyarrow/_dataset.pyx:
```diff
@@ -2746,6 +2746,10 @@ cdef class Scanner(_Weakrefable):
+    use_async : bool, default False
+        If enabled, the an async scanner will be used that should offer
+        better performance with high-latency/highly-parallel filesystems
+        (e.g. S3)
```
The option needs to be added to _populate_builder and Scanner.from_fragment/Scanner.from_dataset or else it won't actually take effect.

lidavidm replied:
Should we also add the parameter to the tests? Maybe some refactoring is needed though to make it easier to share parameters like this.

westonpace replied:
I updated all the tests so that they are parameterized on use_threads/use_async.  This helped identify the spots in _populate_builder, etc. where I had missed passing the flag through.

westonpace replied:
I do think we might want to reduce the number of places we pass parameters around but this was a problem before use_async (e.g. there are a lot of places in the docs where we say "See scan method parameters documentation.").  That's not for this PR though and style-wise it might not be so bad so python can just forward kwargs but when it dips into cython it seems we need to materialized all the kwargs.

lidavidm replied:
It's because `_populate_builder` is cdef, i.e. gets compiled into an actual C function with a C parameter list, so it can't take `**kwargs` (but can directly take things like a shared_ptr). Though, we could pack all the non-C(++) arguments into a single `kwargs`.

westonpace replied:
> Though, we could pack all the non-C(++) arguments into a single kwargs.

I don't know what you mean by this.  Do you mean taking all these arguments and putting them into some kind of `ScanOptions` class?

lidavidm replied:
Ah, sorry, what I mean is that if you're concerned about expanding all the parameters everywhere, you could declare 

```
cdef _populate_builder(shared_ptr[CScannerBuilder]& builder, object kwargs):
    # Extract args from kwargs individually...
```

and then make every user-facing method take just `**kwargs`. Of course, that just pushes the boilerplate around a bit (e.g.potentially having to cast things inside `_populate_builder`).

Doesn't need to be done here, just wanted to mention one way to reduce the number of places where we list all the parameters.

lidavidm commented on r/R/dataset-scan.R:
```diff
@@ -183,6 +191,10 @@ ScannerBuilder <- R6Class("ScannerBuilder", inherit = ArrowObject,
       dataset___ScannerBuilder__UseThreads(self, threads)
       self
     },
+    UseAsync = function(use_async = FALSE) {
```
Did you mean to default to TRUE?

lidavidm replied:
(Otherwise the parameter above doesn't work.)

westonpace replied:
Oops, good catch

westonpace commented:
This PR could use some advice from the R community.  I'm adding the ability to request async (at the moment async is a performance degredation in some cases when I/O is really fast so until we've made more progress there it will need to be optional)  I've added `UseAsync` to the scanner in R which is used, for example, like this:

```
test_that("Scanner$ScanBatches", {
  ds <- open_dataset(ipc_dir, format = "feather")
  batches <- ds$NewScan()$Finish()$ScanBatches()
  table <- Table$create(!!!batches)
  expect_equivalent(as.data.frame(table), rbind(df1, df2))

  batches <- ds$NewScan()$UseAsync(TRUE)$Finish()$ScanBatches()
  table <- Table$create(!!!batches)
  expect_equivalent(as.data.frame(table), rbind(df1, df2))
})
```
However, most of the examples I see reading a dataset are doing something like...

```
ds %>%
      select(string = chr, integer = int) %>%
      filter(integer > 6 & integer < 11) %>%
      collect() %>%
      summarize(mean = mean(integer))
```

How should `UseAsync` be inserted into such a pattern (chain?) of calls.  Should it be it's own operator:

```
ds %>%
      select(string = chr, integer = int) %>%
      filter(integer > 6 & integer < 11) %>%
      use_async() %>%
      collect() %>%
      summarize(mean = mean(integer))
```

...or an argument to `collect`:

```
ds %>%
      select(string = chr, integer = int) %>%
      filter(integer > 6 & integer < 11) %>%
      collect(use_async=TRUE) %>%
      summarize(mean = mean(integer))
```
...or exposed some other different way?

nealrichardson commented:
Elsewhere, we use a global `option` for governing e.g. whether to use multithreading, would that be appropriate here? 

Do you expect this to always be optional, or is it only optional for now while you sort out the performance optimization?

lidavidm commented on python/pyarrow/_dataset.pyx:
```diff
@@ -2969,6 +2982,8 @@ def _filesystemdataset_write(
     Partitioning partitioning not None,
     FileWriteOptions file_options not None,
+    bint use_threads,
+    bint use_async,
```
This doesn't seem used here? Though I guess even before use_threads was ignored.

lidavidm replied:
I guess we'd have to do something like embed ScanOptions into FileSystemDatasetWriteOptions if we want to make these useful, and you already made a followup.

westonpace replied:
Nope, you're right initially.  If the function is already taking in a scanner there is no need for these properties.  They should be specified when building the scanner.

lidavidm commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -206,6 +206,63 @@ def dataset(mockfs):
+            self._patch_kwargs(kwargs)
+            return dataset.scanner(**kwargs)
+
+        def head(self, dataset, num_rows, **kwargs):
```
Note you could declare all these as `(self, dataset, *args, **kwargs)` and avoid having to replicate their individual signatures.

lidavidm replied:
Thanks for converting this file, by the way, this looks like it was rather tedious.

westonpace replied:
I considered this but felt this was more self documenting.  I don't have a strong opinion though, so if you want I'm happy to switch it.

I also realized I could just monkeypatch the dataset/fragment but most tests create their dataset/fragment outside of pytest so that would end up being just as much of a change.

lidavidm replied:
I think the way it is is fine too.

lidavidm commented on python/pyarrow/_dataset.pyx:
```diff
@@ -393,6 +393,11 @@ cdef class Dataset(_Weakrefable):
+            If enabled, the an async scanner will be used that should offer
+            better performance with high-latency/highly-parallel filesystems
+            (e.g. S3)
+
```
Do we want to call it out as still experimental? (Ditto for R below)

westonpace replied:
In python the entire `pyarrow.datasets` module is still experimental.  I don't see much benefit in calling this out specifically.   Given that the API of the asynchronous scanner is the same as the synchronous scanner  I don't expect it to be any more prone to change than what was there before.

lidavidm commented on r/R/dataset-scan.R:
```diff
@@ -42,6 +44,7 @@
 #' - `$UseThreads(threads)`: logical: should the scan use multithreading?
 #' The method's default input is `TRUE`, but you must call the method to enable
 #' multithreading because the scanner default is `FALSE`.
+#' - `$UseAsync(use_async)`: logical: should the async scanner be used?
```
I think you have to roxygenize() to get these changes reflected in the actual docs? (and/or you can just `@github-actions autotune` to have CI do that for you)

westonpace replied:
The changes appear to be reflected in `man/Scanner.Rd` now.  Is that sufficient?  Or should I run autotune?

lidavidm replied:
I think this is good now, thanks. (I confess I'm not exactly sure what R needs.)

lidavidm reviewed (commented):
Looks good to me. Left comments on some small things.

jorisvandenbossche commented on python/pyarrow/_dataset.pyx:
```diff
@@ -393,6 +393,11 @@ cdef class Dataset(_Weakrefable):
             If enabled, then maximum parallelism will be used determined by
             the number of available CPU cores.
+        use_async : bool, default False
+            If enabled, the an async scanner will be used that should offer
```
```suggestion
            If enabled, an async scanner will be used that should offer
```

jorisvandenbossche commented on python/pyarrow/_dataset.pyx:
```diff
@@ -2721,6 +2727,10 @@ cdef class Scanner(_Weakrefable):
         If enabled, then maximum parallelism will be used determined by
         the number of available CPU cores.
+    use_async : bool, default False
+        If enabled, the an async scanner will be used that should offer
```
```suggestion
        If enabled, an async scanner will be used that should offer
```

jorisvandenbossche commented on python/pyarrow/dataset.py:
```diff
@@ -725,6 +725,10 @@ def write_dataset(data, base_dir, basename_template=None, format=None,
         Write files in parallel. If enabled, then maximum parallelism will be
         used determined by the number of available CPU cores.
+    use_async : bool, default False
+        If enabled, the an async scanner will be used that should offer
```
```suggestion
        If enabled, an async scanner will be used that should offer
```

jorisvandenbossche commented:
One concern I had when seeing the heavy parameterization (basically a lot of tests x4) is about the run test time, but checking the most expensive tests, it's basically all related to the S3 related tests:

```
$ pytest python/pyarrow/tests/test_dataset.py --durations=20
...
============================================================================================== slowest 20 durations ===============================================================================================
15.39s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_s3_with_filesystem_uri[threaded-sync]
15.36s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_s3_with_filesystem_uri[serial-sync]
15.36s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_s3_with_filesystem_uri[serial-async]
15.35s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_s3_with_filesystem_uri[threaded-async]
3.13s call     pyarrow/tests/test_dataset.py::test_write_dataset_s3
2.03s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[threaded-async]
2.02s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3_fsspec[threaded-async]
2.02s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[serial-async]
2.02s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3_fsspec[serial-async]
2.02s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[threaded-sync]
2.02s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[serial-sync]
2.02s setup    pyarrow/tests/test_dataset.py::test_write_dataset_s3
2.01s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3_fsspec[serial-sync]
1.58s setup    pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3_fsspec[threaded-sync]
1.07s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[threaded-async]
1.06s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[threaded-sync]
1.05s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[serial-sync]
1.05s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3[serial-async]
0.47s call     pyarrow/tests/test_dataset.py::test_open_dataset_from_uri_s3_fsspec[threaded-async]
0.30s setup    pyarrow/tests/test_dataset.py::test_make_fragment
```

So we now run those 4 times, but instead of objecting to the parameterization (since I suppose it's especially useful for S3 tests?), it's probably more useful to see why it actually takes such a long time

lidavidm commented:
Locally it's far faster with `AWS_EC2_METADATA_DISABLED=true` and test_fs.py already has a workaround for this. (AWS SDK tries to contact an AWS metadata server by default with a long timeout.)

I filed ARROW-12810.

lidavidm commented:
See #10348 - this should speed things up quite a bit (especially in test_fs.py it turns out).

westonpace commented:
> Elsewhere, we use a global option for governing e.g. whether to use multithreading, would that be appropriate here?

A global option seems like it should be fine.

> Do you expect this to always be optional, or is it only optional for now while you sort out the performance optimization?

I expect it should go away at some point.  Either we get to the point where the overhead is negligible or the filesystems themselves could determine what kind of mode to use (potentially via io context).

nealrichardson commented:
👍 then check out examples where `GetBoolOption` is called in `r/src`, that's probably what you want. The option is set in R like `options(name=value)`

westonpace commented:
> +1 then check out examples where GetBoolOption is called in r/src, that's probably what you want. The option is set in R like options(name=value)
@nealrichardson 

I've added a check to use the option if it isn't specified so that should catch dplyr now.  I'm a bit torn on testing.  There is no good way to tell if a scan happened synchronously or asynchronously by looking at the result (ideally they should be identical).  I could add an `IsAsync` method to the scanner but I don't know how I feel about adding that just for testing purposes, but I could probably be swayed.  The proof is really in the performance (and I hope to get an R version of the dataset benchmark up which should demonstrate this).

lidavidm commented on python/pyarrow/dataset.py:
```diff
@@ -780,5 +784,5 @@ def write_dataset(data, base_dir, basename_template=None, format=None,
     _filesystemdataset_write(
         scanner, base_dir, basename_template, filesystem, partitioning,
-        file_options, max_partitions
+        file_options, use_threads, use_async, max_partitions
```
If I'm not mistaken this'll be an error since those parameters don't exist?

lidavidm replied:
They should get passed into L744/L777 above, most likely.

westonpace replied:
Just fixed.

westonpace replied:
Hmm, `use_threads` wasn't being passed in on 744.  I don't think there's any need to pass either since it's just scanning in memory.  I'm not entirely certain on `use_threads` though.

westonpace commented:
> So we now run those 4 times, but instead of objecting to the parameterization (since I suppose it's especially useful for S3 tests?), it's probably more useful to see why it actually takes such a long time
@jorisvandenbossche 

I don't know that the parameterization is particularly useful as the S3 tests are more focused on dataset factory.   I backed off a bit and only one S3 test is now run with async.  This, combined with David's changes, has all tests running in ~2.5s on my desktop.

westonpace commented:
Ok, I think I've addressed all review comments.  Any follow-up is welcome.

However, it might be convenient to hold off on merge for just a bit.  I've got a conbench benchmark that I've added (https://github.com/ursacomputing/benchmarks/pull/40) which should demonstrate that something is actually happening when we set `use_async` to true.  So I'd like to wait until that is merged and I run conbench on the PR.

westonpace commented:
@ursabot please benchmark lang=Python

westonpace commented:
@ursabot please benchmark lang=Python

westonpace commented:
The EC2 instances where we really want the benchmarks are running out of memory it seems.  I was able to get some times from ursa-i9-9960x.  Note, it does not line up the benchmarks because the tags are different but they are the same test.

![image](https://user-images.githubusercontent.com/1696093/118918225-247ede00-b8ce-11eb-8d7e-3914a454ca80.png)

lidavidm commented:
Hmm, seems pre_buffer is a hindrance here :( Doesn't quite line up with the testing I did but that was with a different dataset. We can improve things as a follow up though/I can dig more.

westonpace commented:
It is noise.  The `pre_buffer` option is not actually used in the benchmark if the source is IPC (does IPC even support pre-buffering?).  I should investigate to see if I can remove it in this case.

lidavidm commented:
Ah: it _can_ be used, if you so desire; scan with a custom IpcFragmentScanOptions:

https://github.com/apache/arrow/commit/b01bcf2db2d493b25185be1d06223ccdd1cf33d0#diff-7a6684d2d6b7eae524ff4d645d81616ebc53c726e74bb438150d32136e4a4d83R83-R85

But this isn't exposed to Python.

lidavidm commented:
@westonpace was there any more benchmarking you wanted to do here? Otherwise let's merge this.

westonpace commented:
@lidavidm Go for it :smile:
